### PR TITLE
[e2e tests] Do not usr SRST in OpenOCD reset configuration.

### LINKED ISSUE #17729: [test-triage,rom_e2e,opentitanlib] gdb/openocd/jtag tests flaky/failing
### Hierarchy of regression failure

Chip Level

### Failure Description

The ROM E2E tests that use the `rv_dm + openocd + gdb` are shaky lately.
image

The reason should be investigated, and the tests should be improved.

### Steps to Reproduce

- GitHub Revision: master
- bazel test --define DISABLE_VERILATOR_BUILD=true --define bitstream=gcp_splice --test_tag_filters=-verilator,-dv,-broken --build_tests_only //sw/device/silicon_creator/rom/e2e:rom_e2e_debug_test_otp_test_unlocked0

### Tests with similar or related failures

- [ ] rom_e2e_debug_test_otp_test_unlocked0
- [ ] debug_disallowed_in_prod_fpga_cw310_test_otp_rma
- [ ] rom_e2e_asm_watchdog_bark_fpga_cw310_test_otp_rma
- [ ] rom_e2e_asm_watchdog_bite_fpga_cw310_test_otp_test_unlocked0
- [ ] rom_e2e_asm_watchdog_bite_fpga_cw310_test_otp_dev
- [ ] sram_program_fpga_cw310_test_otp_test_unlocked0

### PR DESCRIPTION
**Pull request by pamaury**
Andreas has identified that OpenOCD was configured to use SRST when issue a `reset`  command. This is problematic since it resets the entire chip, including the JTAG tap and the pinmux. This then causes the pinmux to disable the connection between the pins and the tap a bit later and most likely causes the JTAG issues identified in #17729.
The real chip will not have a srst_n pin and it should not be needed because on RISC-V, issuing a DMI reset though ndmreset will reset *everything* but the part necessary to keep the jtag tap alive, which is exactly what we want.

Hopefully fixes #17729.

### CODE REVIEW

**Comment by johngt** [comment]
@moidx / @msfschaffner for visibility

**Comment by andreaskurth** [comment]
> The real chip will not have a srst_n pin and it should not be needed because on RISC-V, issuing a DMI reset though ndmreset will reset everything but the part necessary to keep the jtag tap alive, which is exactly what we want.

ndmreset does reset a lot, but it's not equivalent to a POR or a SW-requested reset. Especially because ndmreset does *not* interrupt JTAG connectivity, it cannot be used for manufacturing SW that requires LC transitions! CC @msfschaffner

File: rules/scripts/gdb_test_coordinator.py
Comment (andreaskurth) [inline rules/scripts/gdb_test_coordinator.py:162]: I agree that `srst` should not be used because it's only available on the FPGA. Note that we don't have `trst` connected, although it can be emulated with suitable `tms` and `tck` driving for the purpose of resetting the TAP controller.

Comment (andreaskurth) [reply rules/scripts/gdb_test_coordinator.py:162]: Different question: Does OpenOCD / GDB need a way to do a system reset? If so, how will that happen after this change?

Comment (pamaury) [reply rules/scripts/gdb_test_coordinator.py:162]: The GDB tests do not need a system reset. The only tests that require a system resset use opentitanlib that correctly resets the chip using a gpio line (e.g. for LC transitions).

**Comment by pamaury** [comment]
I don't know what `//sw/device/tests:gpio_smoketest_fpga_cw310_test_rom` does, I need to have a look.

**Comment by marnovandermaas** [comment]
> I don't know what `//sw/device/tests:gpio_smoketest_fpga_cw310_test_rom` does, I need to have a look.

I just saw that this is mentioned in Slack, maybe try rebasing it on to the latest main branch.

**Comment by pamaury** [comment]
I have done a lot more testing and it seems to fix the issue. There is one instance where I observed that the one second delay between starting opentitantool and openocd was too small and the console still interefered. Unfortunately, we do have a nice mecanism to report the gpio pins being set. I will check if I can rely on the opentitantool output to block until it is ready.

Longer term, I think it is worth thinking about whether opentitantool should touch the CW310 pins at all when invoked for the console only.

**Comment by timothytrippel** [comment]
> Longer term, I think it is worth thinking about whether opentitantool should touch the CW310 pins at all when invoked for the console only.

Long term, as a lot of us have discussed, I believe we should remove the gdb coordinator script and the custom gdb/openocd Bazel test rule and move the GDB coordination into opentitantool, since we don't need it for writing custom tests now that we drive OpenOCD with opentitanlib directly. I think the only thing we want GDB coordination for is to spin up an interactive debug console, so when we do that integration we can we can tweak opentitantool behavior.

In the interim, first step would be to add the delay between starting opentitantool and openocd as you mention to get the few `opentitan_gdb_fpga_cw310_test` tests working consistently, and the second step would be to transition these tests to custom test harnesses that use opentitanlib to drive openocd directly and remove the opentitan_gdb_fpga_cw310_test rule from the codebase.

Then lastly, is integrate gdb coordination into opentitantool as described above.

**Review by timothytrippel — APPROVED** [review]

**Comment by timothytrippel** [comment]
looks like you just need to fix some lint warning @pamaury , otherwise this is great, thanks for much for all the hard work on this!

**Review by marnovandermaas — APPROVED** [review]
Let's wait for CI to pass before merging.

**Pull request merged**
